2018 Robot/src/org/usfirst/frc/team4999/commands/autonomous/MoveDistance.java

```diff
@@ -0,0 +1,141 @@
+package org.usfirst.frc.team4999.commands.autonomous;
+
+import org.usfirst.frc.team4999.robot.MoPrefs;
+import org.usfirst.frc.team4999.robot.Robot;
+import org.usfirst.frc.team4999.robot.RobotMap;
+import org.usfirst.frc.team4999.robot.sensors.ADIS16448_IMU;
+import org.usfirst.frc.team4999.robot.sensors.VMXPi;
+import org.usfirst.frc.team4999.robot.subsystems.DriveSystem;
+
+import edu.wpi.first.wpilibj.Encoder;
+import edu.wpi.first.wpilibj.PIDController;
+import edu.wpi.first.wpilibj.PIDOutput;
+import edu.wpi.first.wpilibj.PIDSource;
+import edu.wpi.first.wpilibj.PIDSourceType;
+import edu.wpi.first.wpilibj.command.Command;
+
+/**
+ *
+ */
+public class MoveDistance extends Command {
+	
+	private PIDController pid;
+	private Encoder left = RobotMap.leftDriveEncoder;
+	private Encoder right = RobotMap.rightDriveEncoder;
+	private double distance;
+	private static double ticksPerMeter = 1;
+	
+	
+	static class AverageEncoder implements PIDSource{
+		private Encoder left, right;
+
+		private PIDSourceType sourcetype = PIDSourceType.kDisplacement;
+
+
+		
+		public AverageEncoder(Encoder left, Encoder right) {
+			this.left = left;
+			this.right = right;
+		}
+		
+		@Override
+		public void setPIDSourceType(PIDSourceType pidSource) {
+			sourcetype = pidSource;
+		}
+
+		@Override
+		public PIDSourceType getPIDSourceType() {
+			return sourcetype;
+		}
+
+		@Override
+		public double pidGet() {
+			switch(sourcetype) {
+			case kRate:
+				return (left.getRate() + right.getRate()) / 2;
+			case kDisplacement:
+			default:
+				return (left.getDistance() + right.getDistance()) / 2;
+			}
+		}
+		
+	}	
+	
+
+	static class DriveAngleCorrect implements PIDOutput {
+		private DriveSystem output;
+		private Encoder left, right;
+		private double lStart, rStart;
+		VMXPi vmx = VMXPi.getInstance();
+		ADIS16448_IMU adis = new ADIS16448_IMU();
+		
+		private final double distanceBetweenWheels = 0.5; // meters
+		private final double moveErrGain = 0.1;
+		
+		public DriveAngleCorrect(DriveSystem output, Encoder left, Encoder right) {
+			this.output = output;
+			this.left = left;
+			this.right = right;
+			this.lStart = left.getDistance();
+			this.rStart = right.getDistance();
+		}
+		
+		 private double getAngle() { 
+			 if(vmx.getTimeSinceLastPacket() > vmx.RESPONSECUTOFF) {
+				return adis.getAngleZ();
```

Comment:
If you want to fallback to the adis if you lose VMX comms, you're currently assuming that the VMX angle and the ADIS angle correlate and are interchangeable.

To be safe, you should be reading both the ADIS angle and the VMX angle all the time and keeping track of the offset.  Then if the VMX fails, you can produce a number that looks like the VMX by offsetting the ADIS by the last known good offset.